### PR TITLE
The `node.js` binding v0.1.

### PR DESCRIPTION
**Pull request by dkorolev**
image

### CODE REVIEW

File: integrations/nodejs/README.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+* Your project would need to have `node-addon-api` and `bindings` installed via `npm`.
+* Please refer to `current/examples/nodejs` for the reference `package*.json`, `binding.gyp`, and the command lines, in the `Makefile` there.
+
+### Implementation Details
+
+#### Headers
+
+In this directory the header files are of `.hpp` extension. This is primarily to make sure they are excluded from Current's top-level `make check` target.
+
+Also node that these header files are in C++14, even though Current itself is C++17. I was having troubles building C++17 with `node-gyp` cross-platform (things vary across `node-gyp`'s build scripts on Linus and macOS, so I decided to stick with C++14 for now, to be safe).
```
Comment (mzhurovich) [inline integrations/nodejs/README.md]: s/Linus/Linux/ :)

File: integrations/nodejs/README.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+If the C++ function is synchronous, i.e. it does work in one thread and then returns the value, then there is no need to worry about the validity of V8 environment, as the integration code takes care of this.
+
+For asynchronous work, take a look at `JSAsync`. It takes two lambdas as the two arguments: the first lambda is a C++-only one, and the second one is called from within a valid V8 environment.
+
+Thus, the C++ heavylifting can be done within the first lambda, and JavaScript functions can be called, and JavaScript promises can be resolved, from within the second one.
+
+Also, in addition to `JSFunction` (which is `JSFunctionReturning<void>`), this integration also provides the `JSScopedFunction = JSScopedFunctionReturning<void>` primitive, for C++ objects corresponding to C++ functions that do not need to cross the sync/async boundary. A `JSFunction` can be used instead of a `JSScopedFunction` anywhere, while, in async contexts, `JSFunction`, not `JSScopedFunction`, shoul dbe used. In simple terms, `JSScopedFunction` does not indicate to the V8 engine that the function may need to be invoked later, and thus it would not have the async call context created for itself.
```
Comment (mzhurovich) [inline integrations/nodejs/README.md]: s/shoul dbe/should be/

File: integrations/nodejs/README.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+In the "user space" `Init` code, just pass `exports["myFunctionName"] = CPP2JS([](...) { ... });`, and the framework would automatically take care of converting the arguments back and forth.
+
+In addition to the "primitive" types (numbers, strings, `null`-s that are mapped to `nullptr`-s in C++), the instrumentation also handles `JSFunction`-s and `JSPromise`-s correctly. The C++ code can natively call, via `operator()`, the `JSFunction`-s that are passed as parameters from JavaScript. All the arguments would be automatically converted from their C++ representations to the JavaScript ones. Use `JSFunctionReturning<T>`, as the argument type for the lambda, to have the return value of the JavaScript function converted to the respective C++ type.
+
+Of course, as JavaScript is not a strongly typed language and the integration code can not guarantee the user would pass in the arguments of the right type, such an attempt would result in a runtime error.
+
+#### More
+
+Here's a non-exhaustive list of other feature that are either done but not documented yet, or work in progress as of this very moment:
```
Comment (mzhurovich) [inline integrations/nodejs/README.md]: featureS

File: integrations/nodejs/javascript_async_eventloop.hpp [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+namespace javascript {
+namespace impl {
+
+class JSAsyncEventLoopImpl final {
+ private:
+  mutable std::mutex mutex_;
+  mutable std::condition_variable cv_;
+
+  std::function<bool()> next_function_;
+  bool last_call_result_ = true;  bool terminated_ = false;
```
Comment (mzhurovich) [inline integrations/nodejs/javascript_async_eventloop.hpp]: Missing \n

**Review by mzhurovich — APPROVED** [review]
LGTM! Just a few typos

File: integrations/nodejs/javascript_async_eventloop.hpp (hunk 4 above)
Comment (dkorolev) [reply integrations/nodejs/javascript_async_eventloop.hpp]: Oh, I must have introduced it during my final manual review cycle, as `make indent` fixed it. Good eye! :+1:

**Pull request merged**